add examples of usage of GeoParams in basic PT codes

tduretz opened the pull request:
We can store some example of usage of GeoParams together with simple PT codes.
So far only VE loading and inclusion VE loading work.
The codes are unstable and they need an improved damping strategy. I will take care of that soon.
Would be great to invite @boriskaus to this repository such that he can follow and contribute.
Once we get all basic scripts work (including future VEVP and power-law VE runs) we can think of moving the 
scripts to a more appropriate repo (GeoParams?).

Commit:
add examples of usage of GeoPrams in basic PT codes

boriskaus commented:
Sounds good - as to putting this back to the `GeoParams` repo: I would prefer to have a separate repository and leave GeoParams strictly for the point wise computations. Note that in `JustRelax` we are collecting pseudotransient solvers (which should work with `ParallelStencil` and `GeoParams`). Perhaps an idea to store some solvers there that don’t rely on external packages like PS?

Commit:
improved damping stategy

tduretz commented:
Great, I've pushed VE benchmark and VE inclusion with improved damping strategy. They should be much better basis for VEVP implementations. They now include design of PT parameters based on Räss et al. (2022). The damping is applied on the residual rather than on flux components. 
Let's keep these examples in a separate repository, which we could somehow link to Julia Geodynamics.

Commit:
Basic 2D discontinuous Poisson solver that could receive a non-linear conductivity update using GeoParams(?)

boriskaus reviewed (commented):
That should work

luraess commented:
Thanks @tduretz for the contribution! I invited @boriskaus to the repo as well.
I agree that we could at some point move these example to a separate repo that may differ from GeoParams (unless we'd like to have a GD example in there). TBD.

luraess reviewed (changes requested):
@tduretz can you add a `.gitignore` and include in there `.DS_Store` and `.vscode`? 

Also, you may want to have to `scripts` in a scripts folder to keep the `src` for potentially have the main module if needed at some point in there (e.g. in case of adding unit/reference tests).

Commit:
clean up

Commit:
add scripts

Commit:
remove.vscode

Commit:
remove src

boriskaus committed:
added example with plasticity

boriskaus committed:
plot strain rate invariant

boriskaus committed:
added P-dependent plasticity

boriskaus committed:
example using Thibaults parameters (not working)

Commit:
back to visco-elasticity: the visco-elastic inclusion test with GeoParams=false/true gives different results

albert-de-montserrat committed:
fix GeoParams stress update

boriskaus committed:
Merge pull request #4 from albert-de-montserrat/adm-fix-geoparams

fix GeoParams stress update

Commit:
add Eii computation, still weird behaviour

albert-de-montserrat commented:
This is something that we still need (I think) to add to the rheology kernels, but you should not do  `v   = MatParam[Phasec[i]].CompositeRheology[1]` as it is type unstable. This should be replace by ` phase_viscosity(MatParam, ε̇iic[i], Phasec[i], args)` where 
```julia
@generated function phase_viscosity(v::NTuple{N,Any}, ε̇ii, phase, args) where N
    quote
        Base.@_inline_meta
        Base.@nexprs $N i -> v[i].Phase === phase && return computeViscosity_εII(v[i].CompositeRheology[1], ε̇ii, args)        
    end
end
```
This results in about x3.5 speedup in the GeoParams version. I will address this issue in GeoParams when I have a spare minute.

Commit:
Add Albert fix start with timing comparison

boriskaus committed:
example with (regularised) Drucker-Prager plasticity used through GeoParams

tduretz reviewed (commented):
I now get:
```ERROR: UndefVarError: DruckerPrager_regularised not defined```
Should this be defined somewhere in the script or is it a new GP functionality?

boriskaus commented:
that's new in GeoParams; you should link the code with `GeoParams#main`, or `GeoParams#adm-cuda-check` if you also want the latest cuda-check turned off. I can also upload my manifest

boriskaus committed:
added Manifest file (so you have the correct version of GeoParams)

boriskaus committed:
the previous comparison between GeoParams & the native implementation was not entirely fair, as the GeoParams version was updating both vertices and centers whereas the native version only updated centers.

This code makes them more similar (and timings are more similar as well). Yet there appears to be a bug in the "native" VEP version as it explodes even in the purely viscoelastic case. I don't see where I go wrong, so save the current effort

boriskaus committed:
added a version that computes plasticity using invariants; times between the GeoParams and native implementation are much closer now & results the same

boriskaus committed:
code clarity; reduce allocations

boriskaus committed:
debugging: compare the values of the "native" and GP versions. Identical for von Mises (after eliminating a pow2 bug)

boriskaus committed:
bug found; helps if we set the same material properties in both cases...

Commit:
One should now restart from the original script: Stokes2D_vep_reg_ctau.jl

Commit:
fresh restart for VEVP

Commit:
50 shades of VEP stag implementations started...

boriskaus committed:
adding another shade & correcting bug

boriskaus committed:
another shade

albert-de-montserrat committed:
centers version with GP (branch #adm-tuples)

albert-de-montserrat committed:
vertex->center shade with GP (bug somewhere)

albert-de-montserrat committed:
fix index bug

albert-de-montserrat committed:
compare against :inv3

boriskaus committed:
update Manifest & required packages

boriskaus committed:
update script to directly use the GeoParams routine to update deviatoric stresses

boriskaus committed:
using dimensional values

boriskaus committed:
added dilatant rheologies as option for GP (not working!)

albert-de-montserrat committed:
inclusion test with damping v3.0 (linear rheology)

albert-de-montserrat committed:
visco-elasticity working with GP

albert-de-montserrat committed:
bugfix & working plasticity

boriskaus committed:
Auto stash before merge of "add-GeoParams-examples" and "origin/add-GeoParams-examples"

albert-de-montserrat committed:
geting closer (stress slightly lower)

Commit:
bulk elasticity v0

Commit:
add v1, should be GP friendly ;)

luraess commented:
In the codes, shouldn't one use updated pressure in momentum equation (as "usual")? That would make the scheme slightly more stable.

Commit:
first version with plastic dilation --- needs relaxation applied to lambda

tduretz commented:
Yes

> In the codes, shouldn't one use updated pressure in momentum equation (as "usual")? That would make the scheme slightly more stable.

Yes, this can help a little bit. I do this in the latest example with dilation.

albert-de-montserrat committed:
GP_v3 now matches _v2

albert-de-montserrat committed:
ParallelStencil version of damp_V3 with GP

albert-de-montserrat committed:
GPU friendly & cosmetic changes

albert-de-montserrat committed:
fix BC kernel

albert-de-montserrat committed:
fix bcs kernels

albert-de-montserrat committed:
fix a couple of bugs

luraess commented:
What is that status here? Could I merge this upon resolving the conflicts?

boriskaus commented:
Hmm, good question. Yes I think it is ok to merge (haven’t run it in a long time, though)

tduretz committed:
Merge branch 'main' into add-GeoParams-examples

tduretz commented:
I have a suggestion. What about we rename the folder `examples`, which would contain several examples of what one can do with GP. We could have subfolders like '1_flow_laws' (the most basic stuff were discussing about earlier this week, potentially notebooks), '2_deformation_maps' and '3_2D_simulations'. We should still identify which codes in this branch is worth keeping and then move them to '/examples/3_2D_simulations'.

luraess commented:
Organising script within distinct examples with respective folders may be a good idea.

boriskaus commented:
Good idea!
